tests/integrationv2/test_renegotiate.py

```diff
@@ -0,0 +1,259 @@
+        # Indicates that the message will not be received by the peer
+        self.ctrl = ctrl_str is not None
+        if ctrl_str:
+            self.data_str = ctrl_str
+        else:
+            self.data_str = mode.upper() + ":" + str(random.getrandbits(8 * 10))
+
+    @staticmethod
+    def data_to_send(messages, mode):
+        data_bytes = [ ]
+        for message in messages:
+            if message.mode is not mode:
+                continue
+            if message.ctrl:
+                data_bytes.append(to_bytes(message.data_str))
+            else:
+                # Openssl processes initial ASCII characters strangely,
+                # but our framework is not good at handling non-ASCII characters due
+                # to inconsistent use of bytes vs decode and str vs encode.
+                # As a workaround, just prepend a throwaway non-ASCII utf-8 character.
+                data_bytes.append(bytes([0xc2, 0xbb]) + to_bytes(message.data_str))
+        return data_bytes
+    
+    @staticmethod    
+    def expected_output(messages, mode):
+        return [ to_bytes(message.data_str) for message in messages if not message.ctrl and message.mode is not mode ]
+    
+    @staticmethod
+    def send_markers(messages, mode):
+        send_markers = [ ]
+        for i in range(0, len(messages)):
+            message = messages[i]
+            if message.mode is not mode:
+                continue
+            elif message.send_marker:
+                send_markers.append(message.send_marker)
+            elif i == 0:
+                # Assume that the first sender is s2n
+                send_markers.append(S2N.get_send_marker())
+            else:
+                previous = messages[i-1]
+                assert(previous.mode is not mode)
+                send_markers.append(previous.data_str)
+        return send_markers
+    
+    @staticmethod    
+    def close_marker(messages):
+        # Assume that the last sender is the server
+        assert(messages[-1].mode is Provider.ServerMode)
+        output = Msg.expected_output(messages, Provider.ClientMode)
+        return to_marker(output[-1])
+    
+    @staticmethod
+    def debug(messages):
+        print(f'client data to send: {Msg.data_to_send(messages, Provider.ClientMode)}')
+        print(f'server data to send: {Msg.data_to_send(messages, Provider.ServerMode)}')
+        print(f'client send markers: {Msg.send_markers(messages, Provider.ClientMode)}')
+        print(f'server send markers: {Msg.send_markers(messages, Provider.ServerMode)}')
+        print(f'client close_marker: {Msg.close_marker(messages)}')
+        print(f'client expected output: {Msg.expected_output(messages, Provider.ClientMode)}')
+        print(f'server expected output: {Msg.expected_output(messages, Provider.ServerMode)}')
+
+
+# The order of messages that will trigger renegotiation
+# and verify sending and receiving continues to work afterwards.
+RENEG_MESSAGES = [
```

Comment:
What do you think about commenting what each line is supposed to do here. It's currently pretty opaque what is going on.